### PR TITLE
DRDT-55: Featured content in homepage (Marquee Plugin)

### PR DESCRIPTION
**Pull request by fariasf**
https://readersdigest.atlassian.net/browse/DRDT-55

Add support for the Marquee functionality.

### CODE REVIEW

**Comment by fariasf** [comment]
Created new staging environment for pr-65. Here are some useful links:

[WordPress Admin](https://pr-65-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-65-drdt.pantheonsite.io/)
[Article](https://pr-65-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-65-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-65-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-65-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2161/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Congrats! Frontend performance tests passed!

Full Lighthouse performance results:

| Page/Branch | pr-65 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [91](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/pr-65/homepage/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [97](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/pr-65/article/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✔ |
| Listicle | [95](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/pr-65/listicle/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/pr-65/page/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/pr-65/hub/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2160/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

**Comment by fariasf** [comment]
Updated staging environment for pr-65. Here are some useful links:

[WordPress Admin](https://pr-65-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-65-drdt.pantheonsite.io/)
[Article](https://pr-65-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-65-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-65-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-65-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2182/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Congrats! Frontend performance tests passed!

Full Lighthouse performance results:

| Page/Branch | pr-65 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [95](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/pr-65/homepage/lighthouse.report.html) | [97](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [98](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/pr-65/article/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✔ |
| Listicle | [98](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/pr-65/listicle/lighthouse.report.html) | [90](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/pr-65/page/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/pr-65/hub/lighthouse.report.html) | [94](https://circleci.com/gh/fariasf/drdt-test/2181/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

File: web/wp-content/plugins/marquee/inc/metaboxes.php [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+			return;
+		}
+
+		// Only save custom meta for the Marquee post type.
+		$post = get_post( $post_id );
+		if ( $post->post_type !== 'marquee' ) {
+			return;
+		}
+
+		// @todo: Remove dependency on WP_Forms_API.
```
Comment (fariasf) [inline web/wp-content/plugins/marquee/inc/metaboxes.php:117]: See https://readersdigest.atlassian.net/browse/DRDT-93

File: web/wp-content/plugins/marquee/marquee.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Plugin Name: Marquee
+Version: 0.2.3
+Description: Create marquees to highlight featured content.
+Author: Facundo Farias
+Author URI: https://facundofarias.com.ar
+Text Domain: marquee
+*/
+
+// Explicitly declare dependencies
+// @todo: Remove dependency on WP_Forms_API.
```
Comment (fariasf) [inline web/wp-content/plugins/marquee/marquee.php:12]: See https://readersdigest.atlassian.net/browse/DRDT-93

File: web/wp-content/themes/bumblebee/index.php [hunk 3]
```diff
@@ -39,31 +39,35 @@
...
-		<?php if ( have_posts() ) : ?>
+		<?php // @todo: This could be a widget. ?>
+		<?php if ( function_exists( 'get_marquee_post' ) ) : ?>
 			<section class="archive-content pure-g">
 				<?php
-				// @todo: Pick from a custom list of posts.
-				// Featured 1.
-				if ( have_posts() ) {
-					the_post();
+				// @todo: The templates should accept a $post argument to avoid overriding globals.
```
Comment (fariasf) [inline web/wp-content/themes/bumblebee/index.php:61]: See https://readersdigest.atlassian.net/browse/DRDT-94

File: web/wp-content/plugins/marquee/inc/metaboxes.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
+<?php
+
+// @todo: Class is used for namespacing only. Use namespaces.
```
Comment (fariasf) [inline web/wp-content/plugins/marquee/inc/metaboxes.php:3]: When a class has no usage other than namespacing its variables and functions, is there a benefit of keeping it? Or should we just use PHP namespaces?

File: web/wp-content/plugins/marquee/inc/metaboxes.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+class Marquee_Metaboxes {
+	const LINK_META            = '_marquee_link';
+	const FEATURED_META        = '_marquee_featured';
+	const FEATURED_META_SLOT_2 = '_marquee_featured_2'; // @todo: Use a single field like `weight` or `order` or `slot`.
```
Comment (fariasf) [inline web/wp-content/plugins/marquee/inc/metaboxes.php:7]: https://readersdigest.atlassian.net/browse/DRDT-95

File: web/wp-content/plugins/marquee/inc/metaboxes.php [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+	const FEATURED_META_SLOT_3 = '_marquee_featured_3';
+	const CONTENT_ID_META      = '_marquee_content_id';
+	const HOMEPAGE_SLOT_SLUG   = 'homepage-marquee-slot-js'; // @todo: Remove once there's a single field.
```
Comment (fariasf) [inline web/wp-content/plugins/marquee/inc/metaboxes.php:10]: See https://readersdigest.atlassian.net/browse/DRDT-95?focusedCommentId=387017&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-387017

File: web/wp-content/plugins/marquee/legacy/legacy.php [hunk 7]
```diff
@@ -0,0 +1,241 @@
+<?php
```
Comment (fariasf) [inline web/wp-content/plugins/marquee/legacy/legacy.php:1]: This code was not needed for CPT, the file is not included. Keeping it in the repo to revisit later.

File: web/wp-content/plugins/marquee/inc/post-type.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	add_action( 'init', 'register_marquee_post_type', 0 );
+
+	/**
+	 * Use a different permalink when specified, or no link at all. Do this
+	 * because marquees are never supposed to have a permalink of their own.
+	 */
+	function marquee_get_permalink( $permalink, $post, $leavename, $sample ) {
+		if ( $post->post_type === 'marquee' ) {
+			// @todo: If Link is not defined, use Post ID?
```
Comment (fariasf) [inline web/wp-content/plugins/marquee/inc/post-type.php:73]: See https://readersdigest.atlassian.net/browse/DRDT-96

File: web/wp-content/themes/bumblebee/index.php [hunk 9]
```diff
@@ -39,31 +39,35 @@
 				]
 			);
 			?>
-			</section>
-
+		</section>
 
-		<?php if ( have_posts() ) : ?>
+		<?php // @todo: This could be a widget. ?>
```
Comment (fariasf) [inline web/wp-content/themes/bumblebee/index.php:57]: See https://readersdigest.atlassian.net/browse/DRDT-97

**Review by ChrisHursty — APPROVED** [review]
This looks good to me @fariasf - some of the points you make are valid, I would prefer, like you, to move back towards WPCS and away from whatever we got tangled up in now.

**Comment by subinviswanathan** [comment]
@fariasf i see the conflict in composer.json file. i looked wp-forms-api.js lgtm

**Comment by fariasf** [comment]
Updated staging environment for pr-65. Here are some useful links:

[WordPress Admin](https://pr-65-drdt.pantheonsite.io/wp/wp-admin)

[Homepage](https://pr-65-drdt.pantheonsite.io/)
[Article](https://pr-65-drdt.pantheonsite.io/jobsite-tips/abandoned-cable-removal/)
[Listicle](https://pr-65-drdt.pantheonsite.io/jobsite-tips/exterior/12-scaffolding-safety-tips-and-handy-hints/)
[Page](https://pr-65-drdt.pantheonsite.io/about-construction-pro-tips/)
[Hub](https://pr-65-drdt.pantheonsite.io/jobsite-tips/)

**Comment by fariasf** [comment]
There are visual regressions! 

View the [full visual regression test report](https://circleci.com/gh/fariasf/drdt-test/2333/artifacts/0/artifacts/backstop_data/html_report/index.html)

**Comment by fariasf** [comment]
Congrats! Frontend performance tests passed!

Full Lighthouse performance results:

| Page/Branch | pr-65 | master | Result |
| --- | ---: | ---: | :---: |
| Homepage | [96](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/pr-65/homepage/lighthouse.report.html) | [96](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/master/homepage/lighthouse.report.html) | ✔ |
| Article | [98](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/pr-65/article/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/master/article/lighthouse.report.html) | ✔ |
| Listicle | [98](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/pr-65/listicle/lighthouse.report.html) | [97](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/master/listicle/lighthouse.report.html) | ✔ |
| Page | [98](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/pr-65/page/lighthouse.report.html) | [98](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/master/page/lighthouse.report.html) | ✔ |
| Hub | [97](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/pr-65/hub/lighthouse.report.html) | [95](https://circleci.com/gh/fariasf/drdt-test/2334/artifacts/0/artifacts/lighthouse_results/master/hub/lighthouse.report.html) | ✔ |


(**Note: Tests are against the _NOADS_ version of the pages!**)

**Review by angelicagligich — APPROVED** [review]

**Pull request merged**